core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -254,18 +255,36 @@ class ReplicaFetcherThread(name: String,
       // We will not include a replica in the fetch request if it should be throttled.
       if (fetchState.isReadyForFetch && !shouldFollowerThrottle(quota, fetchState, topicPartition)) {
         try {
-          val logStartOffset = this.logStartOffset(topicPartition)
-          builder.add(topicPartition, new FetchRequest.PartitionData(
-            fetchState.fetchOffset, logStartOffset, fetchSize, Optional.of(fetchState.currentLeaderEpoch)))
+          // A fetch needs to be used to update the session only if:
+          //   1. fetch offset has changed
+          //   2. log start offset has changed
+          //   3. leader epoch has changed
+          //   4. the fetch was excluded before and now
+          if (fetchState.updateNeeded) {
+            val logStartOffset = this.logStartOffset(topicPartition)
+            builder.add(topicPartition, new FetchRequest.PartitionData(
+              fetchState.fetchOffset, logStartOffset, fetchSize, Optional.of(fetchState.currentLeaderEpoch)))
+          }
         } catch {
           case _: KafkaStorageException =>
             // The replica has already been marked offline due to log directory failure and the original failure should have already been logged.
             // This partition should be removed from ReplicaFetcherThread soon by ReplicaManager.handleLogDirFailure()
             partitionsWithError += topicPartition
         }
+      } else {
+        // Exclude this partition from the upcoming session;
+        // if we have to exclude this fetch because it is delayed, we need to update its state
+        // so that it would always be included when the delay has elapsed
+        builder.remove(topicPartition)
+        if (fetchState.isDelayed) {
+          markPartitionsForUpdates(topicPartition, fetchState)
+        }
       }
     }
 
+    removedPartitions.foreach { topicPartition => builder.remove(topicPartition) }
+    removedPartitions.clear()
```

Comment:
What will happen if a partition was removed, and then re-added between passes? It's probably unlikely but it seems like it's possible.